File: 프로그래머스/2/131704. 택배상자/택배상자.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(vector<int> order) {
    int answer = 0;
    stack<int> s;
    int cur = 0;
    for(int i=1;i<=order.size();i++) {
        if(i != order[cur]) {
            if(!s.empty() && s.top() == order[cur]){
                s.pop();
                answer++;
                cur++;
                i--;
            }
            else{
                s.push(i);    
            }
        }
        else {
            answer++;
            cur++;
        }
    }

    while(!s.empty()){
        int temp = s.top();
        if(order[cur] == temp){
            answer++;
            cur++;
            s.pop();
        }
        else{
            break;
        }
    }

    return answer;
}

File: 프로그래머스/2/131704. 택배상자/택배상자.py
from collections import deque

def solution(order):
    answer = 0
    order = deque(order)
    sub = []

    for current in range(1, len(order) + 1):
        if len(sub) >= 1:
            if sub[-1] == order[0]:
                order.popleft()
                sub.pop()
                answer += 1

        if current == order[0]:
            order.popleft()
            answer += 1
        else:
            sub.append(current)

    while sub: 
        if not sub.pop() == order.popleft():
            break
        answer += 1
    return answer